File: frontend/src/components/Analytics.js
import React, { useState, useEffect } from 'react';
import {
  Container,
  Paper,
  Typography,
  Box,
  Button,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Chip,
  Alert,
  CircularProgress,
  Tabs,
  Tab,
  Card,
  CardContent,
  Grid,
  FormControl,
  InputLabel,
  Select,
  MenuItem,
  IconButton,
  Dialog,
  DialogTitle,
  DialogContent,
  DialogActions,
  TextField,
  LinearProgress
} from '@mui/material';
import {
  Refresh as RefreshIcon,
  TrendingUp as TrendingUpIcon,
  TrendingDown as TrendingDownIcon,
  Warning as WarningIcon,
  Timeline as TimelineIcon,
  BubbleChart as AnomalyIcon,
  Assessment as AssessmentIcon,
  CheckCircle as CheckCircleIcon,
  Error as ErrorIcon,
  Info as InfoIcon
} from '@mui/icons-material';
import { analyticsAPI } from '../api/api';

function Analytics() {
  const [trends, setTrends] = useState([]);
  const [forecasts, setForecasts] = useState([]);
  const [anomalies, setAnomalies] = useState([]);
  const [summary, setSummary] = useState({
    recent_anomalies: 0,
    average_compliance_7d: 0,
    devices_at_risk: 0,
    total_trends: 0
  });
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);
  const [success, setSuccess] = useState(null);
  const [currentTab, setCurrentTab] = useState(0);
  const [timeRange, setTimeRange] = useState('7d');
  const [openForecastDialog, setOpenForecastDialog] = useState(false);
  const [forecastForm, setForecastForm] = useState({
    device_id: '',
    days_ahead: 7
  });

  useEffect(() => {
    loadData();
  }, [timeRange]);

  const loadData = async () => {
    try {
      setLoading(true);
      const [trendsRes, forecastsRes, anomaliesRes, summaryRes] = await Promise.all([
        analyticsAPI.getTrends({ days: parseInt(timeRange) || 7 }),
        analyticsAPI.getForecast(),
        analyticsAPI.getAnomalies({ acknowledged: false }),
        analyticsAPI.getDashboardSummary()
      ]);
      setTrends(trendsRes.data);
      setForecasts(forecastsRes.data);
      setAnomalies(anomaliesRes.data);
      setSummary(summaryRes.data);
      setError(null);
    } catch (err) {
      setError('Failed to load analytics data: ' + err.message);
    } finally {
      setLoading(false);
    }
  };

  const handleCreateSnapshot = async () => {
    try {
      await analyticsAPI.createSnapshot();
      setSuccess('Snapshot created successfully');
      loadData();
    } catch (err) {
      setError('Failed to create snapshot: ' + err.message);
    }
  };

  const handleGenerateForecast = async () => {
    try {
      const deviceId = forecastForm.device_id ? parseInt(forecastForm.device_id) : null;
      await analyticsAPI.generateForecast(deviceId, forecastForm.days_ahead);
      setSuccess('Forecast generated successfully');
      setOpenForecastDialog(false);
      loadData();
    } catch (err) {
      setError('Failed to generate forecast: ' + err.message);
    }
  };

  const handleDetectAnomalies = async () => {
    try {
      await analyticsAPI.detectAnomalies();
      setSuccess('Anomaly detection completed');
      loadData();
    } catch (err) {
      setError('Failed to detect anomalies: ' + err.message);
    }
  };

  const handleAcknowledgeAnomaly = async (id) => {
    try {
      const user = 'admin'; // In real app, get from auth context
      await analyticsAPI.acknowledgeAnomaly(id, user);
      setSuccess('Anomaly acknowledged');
      loadData();
    } catch (err) {
      setError('Failed to acknowledge anomaly: ' + err.message);
    }
  };

  const getTrendIcon = (change) => {
    if (change > 0) return <TrendingUpIcon color="success" />;
    if (change < 0) return <TrendingDownIcon color="error" />;
    return <TimelineIcon color="action" />;
  };

  const getSeverityColor = (severity) => {
    const colors = {
      critical: 'error',
      high: 'warning',
      medium: 'info',
      low: 'default'
    };
    return colors[severity] || 'default';
  };

  const getComplianceColor = (score) => {
    if (score >= 90) return 'success';
    if (score >= 70) return 'warning';
    return 'error';
  };

  const formatDate = (dateString) => {
    if (!dateString) return 'N/A';
    return new Date(dateString).toLocaleDateString();
  };

  if (loading) {
    return (
      <Box display="flex" justifyContent="center" alignItems="center" minHeight="400px">
        <CircularProgress />
      </Box>
    );
  }

  return (
    <Container maxWidth="lg" sx={{ mt: 4, mb: 4 }}>
      <Box display="flex" justifyContent="space-between" alignItems="center" mb={3}>
        <Typography variant="h4" component="h1">
          Analytics & Forecasting
        </Typography>
        <Box>
          <FormControl size="small" sx={{ minWidth: 120, mr: 1 }}>
            <InputLabel>Time Range</InputLabel>
            <Select
              value={timeRange}
              onChange={(e) => setTimeRange(e.target.value)}
              label="Time Range"
            >
              <MenuItem value="7d">7 Days</MenuItem>
              <MenuItem value="14d">14 Days</MenuItem>
              <MenuItem value="30d">30 Days</MenuItem>
              <MenuItem value="90d">90 Days</MenuItem>
            </Select>
          </FormControl>
          <Button
            variant="outlined"
            startIcon={<RefreshIcon />}
            onClick={loadData}
            sx={{ mr: 1 }}
          >
            Refresh
          </Button>
        </Box>
      </Box>

      {error && (
        <Alert severity="error" onClose={() => setError(null)} sx={{ mb: 2 }}>
          {error}
        </Alert>
      )}

      {success && (
        <Alert severity="success" onClose={() => setSuccess(null)} sx={{ mb: 2 }}>
          {success}
        </Alert>
      )}

      {/* Statistics Cards */}
      <Grid container spacing={3} mb={3}>
        <Grid item xs={12} sm={6} md={3}>
          <Card>
            <CardContent>
              <Box display="flex" alignItems="center" justifyContent="space-between">
                <Box>
                  <Typography color="text.secondary" gutterBottom variant="body2">
                    Avg Compliance (7d)
                  </Typography>
                  <Typography variant="h4" color={getComplianceColor(summary.average_compliance_7d)}>
                    {summary.average_compliance_7d.toFixed(1)}%
                  </Typography>
                </Box>
                <AssessmentIcon sx={{ fontSize: 40, color: 'primary.main' }} />
              </Box>
              <LinearProgress
                variant="determinate"
                value={summary.average_compliance_7d}
                color={getComplianceColor(summary.average_compliance_7d)}
                sx={{ mt: 2, height: 6, borderRadius: 3 }}
              />
            </CardContent>
          </Card>
        </Grid>
        <Grid item xs={12} sm={6} md={3}>
          <Card>
            <CardContent>
              <Box display="flex" alignItems="center" justifyContent="space-between">
                <Box>
                  <Typography color="text.secondary" gutterBottom variant="body2">
                    Recent Anomalies
                  </Typography>
                  <Typography variant="h4" color="warning.main">
                    {summary.recent_anomalies}
                  </Typography>
                </Box>
                <AnomalyIcon sx={{ fontSize: 40, color: 'warning.main' }} />
              </Box>
            </CardContent>
          </Card>
        </Grid>
        <Grid item xs={12} sm={6} md={3}>
          <Card>
            <CardContent>
              <Box display="flex" alignItems="center" justifyContent="space-between">
                <Box>
                  <Typography color="text.secondary" gutterBottom variant="body2">
                    Devices at Risk
                  </Typography>
                  <Typography variant="h4" color="error.main">
                    {summary.devices_at_risk}
                  </Typography>
                </Box>
                <ErrorIcon sx={{ fontSize: 40, color: 'error.main' }} />
              </Box>
            </CardContent>
          </Card>
        </Grid>
        <Grid item xs={12} sm={6} md={3}>
          <Card>
            <CardContent>
              <Box display="flex" alignItems="center" justifyContent="space-between">
                <Box>
                  <Typography color="text.secondary" gutterBottom variant="body2">
                    Total Trends
                  </Typography>
                  <Typography variant="h4">{summary.total_trends}</Typography>
                </Box>
                <TimelineIcon sx={{ fontSize: 40, color: 'info.main' }} />
              </Box>
            </CardContent>
          </Card>
        </Grid>
      </Grid>

      <Paper sx={{ mb: 3 }}>
        <Tabs value={currentTab} onChange={(e, v) => setCurrentTab(v)}>
          <Tab label={`Trends (${trends.length})`} icon={<TimelineIcon />} iconPosition="start" />
          <Tab label={`Forecasts (${forecasts.length})`} icon={<TrendingUpIcon />} iconPosition="start" />
          <Tab label={`Anomalies (${anomalies.length})`} icon={<AnomalyIcon />} iconPosition="start" />
        </Tabs>

        {currentTab === 0 && (
          <Box>
            <Box p={2} display="flex" justifyContent="flex-end">
              <Button
                variant="contained"
                size="small"
                onClick={handleCreateSnapshot}
              >
                Create Snapshot
              </Button>
            </Box>
            <TableContainer>
              <Table>
                <TableHead>
                  <TableRow>
                    <TableCell>Date</TableCell>
                    <TableCell>Overall Compliance</TableCell>
                    <TableCell>Change</TableCell>
                    <TableCell>Total Devices</TableCell>
                    <TableCell>Compliant Devices</TableCell>
                    <TableCell>Failed Devices</TableCell>
                  </TableRow>
                </TableHead>
                <TableBody>
                  {trends.length === 0 ? (
                    <TableRow>
                      <TableCell colSpan={6} align="center">
                        <Typography variant="body2" color="text.secondary" py={3}>
                          No trend data available. Create snapshots to track compliance over time.
                        </Typography>
                      </TableCell>
                    </TableRow>
                  ) : (
                    trends.map((trend, idx) => (
                      <TableRow key={idx}>
                        <TableCell>{formatDate(trend.snapshot_date)}</TableCell>
                        <TableCell>
                          <Box display="flex" alignItems="center">
                            <Typography variant="body2" fontWeight="medium" mr={1}>
                              {trend.overall_compliance?.toFixed(1)}%
                            </Typography>
                            <LinearProgress
                              variant="determinate"
                              value={trend.overall_compliance || 0}
                              color={getComplianceColor(trend.overall_compliance || 0)}
                              sx={{ width: 100, height: 6, borderRadius: 3 }}
                            />
                          </Box>
                        </TableCell>
                        <TableCell>
                          <Box display="flex" alignItems="center">
                            {getTrendIcon(trend.compliance_change)}
                            <Typography variant="body2" ml={1}>
                              {trend.compliance_change > 0 ? '+' : ''}
                              {trend.compliance_change?.toFixed(1) || '0'}%
                            </Typography>
                          </Box>
                        </TableCell>
                        <TableCell>{trend.total_devices || 0}</TableCell>
                        <TableCell>
                          <Chip
                            label={trend.compliant_devices || 0}
                            size="small"
                            color="success"
                            icon={<CheckCircleIcon />}
                          />
                        </TableCell>
                        <TableCell>
                          <Chip
                            label={trend.failed_devices || 0}
                            size="small"
                            color="error"
                            icon={<ErrorIcon />}
                          />
                        </TableCell>
                      </TableRow>
                    ))
                  )}
                </TableBody>
              </Table>
            </TableContainer>
          </Box>
        )}

        {currentTab === 1 && (
          <Box>
            <Box p={2} display="flex" justifyContent="flex-end">
              <Button
                variant="contained"
                size="small"
                onClick={() => setOpenForecastDialog(true)}
              >
                Generate Forecast
              </Button>
            </Box>
            <TableContainer>
              <Table>
                <TableHead>
                  <TableRow>
                    <TableCell>Forecast Date</TableCell>
                    <TableCell>Predicted Compliance</TableCell>
                    <TableCell>Confidence</TableCell>
                    <TableCell>Device</TableCell>
                    <TableCell>Generated</TableCell>
                  </TableRow>
                </TableHead>
                <TableBody>
                  {forecasts.length === 0 ? (
                    <TableRow>
                      <TableCell colSpan={5} align="center">
                        <Typography variant="body2" color="text.secondary" py={3}>
                          No forecast data available. Generate forecasts to predict future compliance trends.
                        </Typography>
                      </TableCell>
                    </TableRow>
                  ) : (
                    forecasts.map((forecast, idx) => (
                      <TableRow key={idx}>
                        <TableCell>{formatDate(forecast.forecast_date)}</TableCell>
                        <TableCell>
                          <Box display="flex" alignItems="center">
                            <Typography variant="body2" fontWeight="medium" mr={1}>
                              {forecast.predicted_compliance?.toFixed(1)}%
                            </Typography>
                            <LinearProgress
                              variant="determinate"
                              value={forecast.predicted_compliance || 0}
                              color={getComplianceColor(forecast.predicted_compliance || 0)}
                              sx={{ width: 100, height: 6, borderRadius: 3 }}
                            />
                          </Box>
                        </TableCell>
                        <TableCell>
                          <Chip
                            label={`${(forecast.confidence_score * 100)?.toFixed(0)}%`}
                            size="small"
                            color={forecast.confidence_score > 0.7 ? 'success' : 'warning'}
                          />
                        </TableCell>
                        <TableCell>{forecast.device_id || 'All Devices'}</TableCell>
                        <TableCell>{formatDate(forecast.created_at)}</TableCell>
                      </TableRow>
                    ))
                  )}
                </TableBody>
              </Table>
            </TableContainer>
          </Box>
        )}

        {currentTab === 2 && (
          <Box>
            <Box p={2} display="flex" justifyContent="flex-end">
              <Button
                variant="contained"
                size="small"
                onClick={handleDetectAnomalies}
              >
                Run Anomaly Detection
              </Button>
            </Box>
            <TableContainer>
              <Table>
                <TableHead>
                  <TableRow>
                    <TableCell>Type</TableCell>
                    <TableCell>Severity</TableCell>
                    <TableCell>Device</TableCell>
                    <TableCell>Description</TableCell>
                    <TableCell>Z-Score</TableCell>
                    <TableCell>Detected</TableCell>
                    <TableCell>Status</TableCell>
                    <TableCell>Actions</TableCell>
                  </TableRow>
                </TableHead>
                <TableBody>
                  {anomalies.length === 0 ? (
                    <TableRow>
                      <TableCell colSpan={8} align="center">
                        <Typography variant="body2" color="text.secondary" py={3}>
                          No anomalies detected. Run anomaly detection to identify unusual patterns.
                        </Typography>
                      </TableCell>
                    </TableRow>
                  ) : (
                    anomalies.map((anomaly) => (
                      <TableRow key={anomaly.id}>
                        <TableCell>
                          <Chip
                            label={anomaly.anomaly_type}
                            size="small"
                            color="primary"
                            variant="outlined"
                          />
                        </TableCell>
                        <TableCell>
                          <Chip
                            label={anomaly.severity}
                            size="small"
                            color={getSeverityColor(anomaly.severity)}
                            icon={<WarningIcon />}
                          />
                        </TableCell>
                        <TableCell>{anomaly.device_id || 'N/A'}</TableCell>
                        <TableCell>
                          <Typography variant="body2">
                            {anomaly.description || 'No description'}
                          </Typography>
                        </TableCell>
                        <TableCell>
                          {anomaly.z_score ? (
                            <Chip
                              label={anomaly.z_score.toFixed(2)}
                              size="small"
                              color={Math.abs(anomaly.z_score) > 3 ? 'error' : 'warning'}
                            />
                          ) : (
                            'N/A'
                          )}
                        </TableCell>
                        <TableCell>{formatDate(anomaly.detected_at)}</TableCell>
                        <TableCell>
                          {anomaly.acknowledged ? (
                            <Chip
                              label="Acknowledged"
                              size="small"
                              color="success"
                              icon={<CheckCircleIcon />}
                            />
                          ) : (
                            <Chip
                              label="New"
                              size="small"
                              color="warning"
                            />
                          )}
                        </TableCell>
                        <TableCell>
                          {!anomaly.acknowledged && (
                            <Button
                              size="small"
                              variant="outlined"
                              onClick={() => handleAcknowledgeAnomaly(anomaly.id)}
                            >
                              Acknowledge
                            </Button>
                          )}
                        </TableCell>
                      </TableRow>
                    ))
                  )}
                </TableBody>
              </Table>
            </TableContainer>
          </Box>
        )}
      </Paper>

      {/* Generate Forecast Dialog */}
      <Dialog open={openForecastDialog} onClose={() => setOpenForecastDialog(false)} maxWidth="sm" fullWidth>
        <DialogTitle>Generate Compliance Forecast</DialogTitle>
        <DialogContent>
          <TextField
            fullWidth
            label="Device ID (optional)"
            type="number"
            value={forecastForm.device_id}
            onChange={(e) => setForecastForm({ ...forecastForm, device_id: e.target.value })}
            margin="normal"
            helperText="Leave empty to forecast for all devices"
          />

          <FormControl fullWidth margin="normal">
            <InputLabel>Forecast Period</InputLabel>
            <Select
              value={forecastForm.days_ahead}
              onChange={(e) => setForecastForm({ ...forecastForm, days_ahead: e.target.value })}
              label="Forecast Period"
            >
              <MenuItem value={7}>7 days ahead</MenuItem>
              <MenuItem value={14}>14 days ahead</MenuItem>
              <MenuItem value={30}>30 days ahead</MenuItem>
              <MenuItem value={90}>90 days ahead</MenuItem>
            </Select>
          </FormControl>

          <Alert severity="info" sx={{ mt: 2 }}>
            <Typography variant="body2">
              Forecasting uses historical compliance data to predict future trends. More historical data will improve accuracy.
            </Typography>
          </Alert>
        </DialogContent>
        <DialogActions>
          <Button onClick={() => setOpenForecastDialog(false)}>Cancel</Button>
          <Button onClick={handleGenerateForecast} variant="contained">
            Generate
          </Button>
        </DialogActions>
      </Dialog>
    </Container>
  );
}

export default Analytics;

File: services/analytics-service/app/routes/analytics.py
# ============================================================================
# routes/analytics.py - Analytics API
# ============================================================================

from fastapi import APIRouter, HTTPException, Depends, Query, Body
from sqlalchemy.orm import Session
from typing import Optional, List
from datetime import datetime

from shared.database import get_db
from app.services.analytics_service import AnalyticsService

router = APIRouter(prefix="/analytics", tags=["Analytics"])


# ============================================================================
# Compliance Trends
# ============================================================================

@router.get("/trends")
async def get_trends(
    days: int = Query(7, ge=1, le=365),
    device_id: Optional[int] = None,
    db: Session = Depends(get_db)
):
    """
    Get compliance trends over time
    
    Args:
        days: Number of days to look back (1-365)
        device_id: Optional device ID to filter trends
        
    Returns:
        List of trend snapshots with compliance metrics
    """
    try:
        trends = await AnalyticsService.get_trends(db, device_id, days)
        
        return [{
            "id": t.id,
            "snapshot_date": t.snapshot_date.isoformat() if t.snapshot_date else None,
            "device_id": t.device_id,
            "overall_compliance": t.overall_compliance,
            "compliance_change": t.compliance_change,
            "total_devices": t.total_devices,
            "compliant_devices": t.compliant_devices,
            "failed_devices": t.failed_devices,
            "total_checks": t.total_checks,
            "passed_checks": t.passed_checks,
            "failed_checks": t.failed_checks,
            "warning_checks": t.warning_checks,
            "critical_failures": t.critical_failures,
            "high_failures": t.high_failures,
            "medium_failures": t.medium_failures,
            "low_failures": t.low_failures
        } for t in trends]
        
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to fetch trends: {str(e)}"
        )


@router.post("/trends/snapshot")
async def create_snapshot(
    device_id: Optional[int] = Body(None, embed=True),
    db: Session = Depends(get_db)
):
    """
    Create a compliance snapshot for trend tracking
    
    Args:
        device_id: Optional device ID (None for overall snapshot)
        
    Returns:
        Created snapshot
    """
    try:
        snapshot = await AnalyticsService.create_snapshot(db, device_id)
        
        return {
            "id": snapshot.id,
            "snapshot_date": snapshot.snapshot_date.isoformat(),
            "device_id": snapshot.device_id,
            "overall_compliance": snapshot.overall_compliance,
            "compliance_change": snapshot.compliance_change,
            "total_devices": snapshot.total_devices,
            "message": "Snapshot created successfully"
        }
        
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to create snapshot: {str(e)}"
        )


# ============================================================================
# Compliance Forecasting
# ============================================================================

@router.get("/forecast")
async def get_forecast(
    device_id: Optional[int] = None,
    db: Session = Depends(get_db)
):
    """
    Get existing compliance forecasts
    
    Args:
        device_id: Optional device ID to filter forecasts
        
    Returns:
        List of forecast predictions
    """
    try:
        forecasts = await AnalyticsService.get_forecasts(db, device_id)
        
        return [{
            "id": f.id,
            "forecast_date": f.forecast_date.isoformat(),
            "device_id": f.device_id,
            "predicted_compliance": f.predicted_compliance,
            "confidence_score": f.confidence_score,
            "predicted_failures": f.predicted_failures,
            "model_version": f.model_version,
            "training_data_points": f.training_data_points,
            "created_at": f.created_at.isoformat() if f.created_at else None
        } for f in forecasts]
        
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to fetch forecasts: {str(e)}"
        )


@router.post("/forecast/generate")
async def generate_forecast(
    device_id: Optional[int] = Body(None),
    days_ahead: int = Body(7, ge=1, le=90),
    db: Session = Depends(get_db)
):
    """
    Generate new compliance forecasts using historical data
    
    Args:
        device_id: Optional device ID (None for overall forecast)
        days_ahead: Number of days to forecast (1-90)
        
    Returns:
        Generated forecasts
    """
    try:
        forecasts = await AnalyticsService.generate_forecast(db, device_id, days_ahead)
        
        if not forecasts:
            return {
                "message": "Not enough historical data to generate forecast. Need at least 2 trend snapshots.",
                "forecasts": []
            }
        
        return {
            "message": f"Generated {len(forecasts)} forecast(s) for the next {days_ahead} days",
            "forecasts": [{
                "id": f.id,
                "forecast_date": f.forecast_date.isoformat(),
                "predicted_compliance": f.predicted_compliance,
                "confidence_score": f.confidence_score,
                "predicted_failures": f.predicted_failures
            } for f in forecasts]
        }
        
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to generate forecast: {str(e)}"
        )


# ============================================================================
# Anomaly Detection
# ============================================================================

@router.get("/anomalies")
async def get_anomalies(
    acknowledged: Optional[bool] = Query(None),
    device_id: Optional[int] = None,
    db: Session = Depends(get_db)
):
    """
    Get detected compliance anomalies
    
    Args:
        acknowledged: Filter by acknowledgment status
        device_id: Optional device ID to filter anomalies
        
    Returns:
        List of detected anomalies
    """
    try:
        anomalies = await AnalyticsService.get_anomalies(db, device_id, acknowledged)
        
        return [{
            "id": a.id,
            "device_id": a.device_id,
            "detected_at": a.detected_at.isoformat(),
            "anomaly_type": a.anomaly_type,
            "severity": a.severity.value if a.severity else None,
            "description": a.description,
            "z_score": a.z_score,
            "expected_value": a.expected_value,
            "actual_value": a.actual_value,
            "acknowledged": a.acknowledged,
            "acknowledged_by": a.acknowledged_by,
            "acknowledged_at": a.acknowledged_at.isoformat() if a.acknowledged_at else None,
            "resolution_notes": a.resolution_notes
        } for a in anomalies]
        
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to fetch anomalies: {str(e)}"
        )


@router.post("/anomalies/detect")
async def detect_anomalies(
    device_id: Optional[int] = Body(None, embed=True),
    db: Session = Depends(get_db)
):
    """
    Run anomaly detection on compliance data
    
    Args:
        device_id: Optional device ID (None for all devices)
        
    Returns:
        Detected anomalies
    """
    try:
        anomalies = await AnalyticsService.detect_anomalies(db, device_id)
        
        return {
            "message": f"Detected {len(anomalies)} anomalie(s)",
            "anomalies": [{
                "id": a.id,
                "device_id": a.device_id,
                "anomaly_type": a.anomaly_type,
                "severity": a.severity.value if a.severity else None,
                "description": a.description,
                "z_score": a.z_score
            } for a in anomalies]
        }
        
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to detect anomalies: {str(e)}"
        )


@router.post("/anomalies/{anomaly_id}/acknowledge")
async def acknowledge_anomaly(
    anomaly_id: int,
    acknowledged_by: str = Body(...),
    notes: Optional[str] = Body(None),
    db: Session = Depends(get_db)
):
    """
    Acknowledge an anomaly
    
    Args:
        anomaly_id: ID of the anomaly to acknowledge
        acknowledged_by: Username of person acknowledging
        notes: Optional resolution notes
        
    Returns:
        Updated anomaly
    """
    try:
        anomaly = await AnalyticsService.acknowledge_anomaly(
            db, anomaly_id, acknowledged_by, notes
        )
        
        return {
            "message": "Anomaly acknowledged",
            "anomaly": {
                "id": anomaly.id,
                "acknowledged": anomaly.acknowledged,
                "acknowledged_by": anomaly.acknowledged_by,
                "acknowledged_at": anomaly.acknowledged_at.isoformat()
            }
        }
        
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to acknowledge anomaly: {str(e)}"
        )


# ============================================================================
# Dashboard Summary
# ============================================================================

@router.get("/dashboard/summary")
async def get_dashboard_summary(db: Session = Depends(get_db)):
    """
    Get analytics dashboard summary with key metrics
    
    Returns:
        Summary statistics for the analytics dashboard
    """
    try:
        summary = await AnalyticsService.get_dashboard_summary(db)
        return summary
        
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to fetch dashboard summary: {str(e)}"
        )
